File: LeetCode/C++/100-199/155.cpp
#include <iostream>
#include <stack>
using namespace std;

class MinStack {
    stack<int> myStack;
    stack<int> minStack;
public:
    /** initialize your data structure here. */
    MinStack() {

    }

    void push(int x) {
        myStack.push(x);
        if(minStack.empty() || minStack.top() >= x) {
             minStack.push(x);
        }
    }

    void pop() {
        int curVal = myStack.top();
        if(curVal == minStack.top())
            minStack.pop();

        myStack.pop();
    }

    int top() {
        return myStack.top();
    }

    int getMin() {
        return minStack.top();
    }
};

int main()
{
    MinStack myStack;
    myStack.push(1);
    myStack.push(2);
    myStack.push(-1);
    myStack.push(-1);
    cout << myStack.getMin() << endl;
    cout << myStack.top() << endl;
    myStack.push(3);
    myStack.pop();
    myStack.pop();
    cout << myStack.getMin() << endl;
    cout << myStack.top() << endl;
    return 0;
}

File: LeetCode/C#/100-199/155.cs
using System;
using System.Collections.Generic;

namespace LeetCode155
{
    public class MinStack {

        Stack<int> myStack = new Stack<int>();
        Stack<int> minStack = new Stack<int>();
        
        /** initialize your data structure here. */
        public MinStack() {
            
        }
        
        public void Push(int x) {
            myStack.Push(x);
            if(minStack.Count == 0 || minStack.Peek() >= x) {
                minStack.Push(x);
            }
            
        }
        
        public void Pop() {
            int curVal = myStack.Peek();
            if(curVal == minStack.Peek())
                minStack.Pop();
            
            myStack.Pop();
            
        }
        
        public int Top() {
            return myStack.Peek();
            
        }
        
        public int GetMin() {
            return minStack.Peek();
        }
    }


    class Program
    {
        static void Main(string[] args)
        {
            MinStack myStack = new MinStack();
            myStack.Push(1);
            myStack.Push(2);
            myStack.Push(-1);
            myStack.Push(-1);
            Console.WriteLine(myStack.GetMin());
            Console.WriteLine(myStack.Top());
            myStack.Push(3);
            myStack.Pop();
            myStack.Pop();
            Console.WriteLine(myStack.GetMin());
            Console.WriteLine(myStack.Top());
        }
    }
}
